### ISSUE TITLE
Allow multiple test reports

### ISSUE THREAD

**Issue by reddwarf69** [jenkinsci/junit-plugin#432]
### What feature do you want to see added?

Warnings-NG has two parameters: "id" and "name" (https://www.jenkins.io/doc/pipeline/steps/warnings-ng/#recordissues-record-compiler-warnings-and-static-analysis-results) that allow you to have separate reports.

junit has "checksName", that allow you to have separate reports in GitHub... but those are mixed, all together, when you go back to Jenkins.

I would like to have the "id" and "name" options also in junit.

**Comment by timja** [comment]
could you give a bigger example? JUnit does use the 'enclosing block'.

So it will include e.g. the stage name in the final checks name

**Comment by reddwarf69** [comment]
I'm OK with how the Checks API is being used / GitHub Checks are being published.
The problem is in how different sets of tests are presented in the Jenkins UI.

This is the side of my job in the Jenkins UI.
![Jenkins](https://user-images.githubusercontent.com/1608276/190142814-94170e3b-ebaf-446c-9d7e-3cfbbb134505.png)

"clang-tidy Warnings" and "clang-tidy servers Warnings" are both created by Warnings NG, but they are clearly split.

They are also split as GitHub Checks

![Tests](https://user-images.githubusercontent.com/1608276/190143033-ea07094c-1559-4651-954f-ec860b35d804.png)

And yes, "Tests" and "Server Tests" are also split as GitHub Checks. I can even see how many of each tests there are (160 "Tests" and 2 "Server tests")

![Server tests](https://user-images.githubusercontent.com/1608276/190143218-428e828e-2217-4f65-a2a0-cd6f3f01977a.png)

But in the Jenkins UI there is only "Test Result", and if I click on it, I get

![Jenkins Tests](https://user-images.githubusercontent.com/1608276/190143288-fffa7ce3-7f2d-4a2a-b6f2-ae81a4facff4.png)

a single number: 162 test (160+2).

Since they are split as GitHub checks, I would like them split in the Jenkins UI too. And I can do that with the Warnings NG results with its "id" and "name" options.

```
    recordIssues 
                 tools: [
                   groovyScript(parserId: 'my_clangtidy', pattern: 'logs/build_full_static_analysis.log'),
                   groovyScript(parserId: 'my_clangtidy', id: 'my_clangtidy_servers', name: 'clang-tidy servers', pattern: 'logs/build_full_servers_static_analysis.log')
                 ]
```

I can also do it when creating the documentation too
```
  publishHTML([
    ...
    reportName: 'Server Documentation'
  ]
```

But I can't do it with the tests (and the coverage report, but that's a different issue in a different plugin).

**Comment by uhafner** [comment]
This is an interesting point. Recently I thought that it might make more sense to remove the UI part of that feature from the warnings plugin and show only a single "warnings" link in order to remove the clutter. Using the new [application bar](https://github.com/jenkinsci/jenkins/pull/7051) one can start within the top level warnings link and then show the individual subsections of the warnings in separate sub-pages. (The same I am planning in the coverage plugin: one link and summary, but multiple results below this entry point). With the 1-link approach it means we would have a better hierarchical decomposition of the Jenkins builds results. 

I understand that the current JUnit results view is kind of awful and has a poor UI that should get a facelift. So this feature does not work with the current implementation in the JUnit plugin and requires some extra work. What is not yet clear to me: is your feature request about the option to show filtered results using separate sidebar links, or is the actual requirement to implement a persistent filter of the JUnit results that still can be navigated from one single link to a page that has several results available (e.g. using tabs or any other separation)?

**Comment by reddwarf69** [comment]
I guess what I want is to have the results of two independent projects into a single Jenkins job, in a way as close as possible to "just twice what I already had when there was a single project".
But to be fair, the problem may be me having two independent projects not only in a single repository, but also building as a single Jenkins job.

Splitting it into multiple repos is not trivial. But I have actually just created a second Jenkinsfile in the repo, to build the second project. Since I now have two independent Jenkins jobs, I can have my "Tests" and "Server Tests" in the Jenkins UI, each one in a different job; which makes more sense to start with.
Also, it didn't make any sense to me for junit to be able to split the checks in GitHub but not in the Jenkins UI. It was either none or both, but not only one. Now that I have two Jenkins jobs, it makes sense! I need to change the name in one of the jobs to avoid a conflict, with two jobs sending the same "check" to GitHub for the same commit.

So... I was asking this because Warnings NG also had it, and it made sense for all the plugins to behave similarly. But since Warnings NG is going to remove it, and I have a better fix for my workflow, I guess this issue could be closed.

**Comment by timja** [comment]
I think fine to keep open for now.

The test reports should likely be named better / separated. I don't think archiving multiple ones is done terribly well

**Comment by yogeek** [comment]
we would also benefit from a better separation between different test reports in jenkins UI : is this proposition still active please ?

**Comment by kaddkaka** [comment]
We are also looking for having and supporting multiple result files. The reason is that we run our testsuite in a few different configurations and create one report for each run. The number of tests can become quite large (~50 k) which makes the html gui render very slowly.

I created an issue on the Jenkins JIRA but was redirected here. (https://issues.jenkins.io/browse/JENKINS-73594)

> When having a long tests result, the Web GUI is too slow to handle it satisfactory. Therefore, the automatic merging of several junit test result xml files is unwanted behavior.
> 
> I have a pipeline that generates 10 different result files. It would be great if Jenkins could leave them untouched and just present all 10 as separate test results for the same build.
> 
> Similar feature was requested for different purpose here: https://stackoverflow.com/questions/41871285/using-multiple-junit-results-file-on-jenkins-pipeline

**Comment by canbuyukburc** [comment]
we are also making different tests for different platforms so it would be better to have test results seperately.
Like in publishHTML plugin a reportName attribute would help a lot for different teams to see their report.

**Comment by boardbloke** [comment]
We would also like this

**Comment by st0yanDimitrov** [comment]
+1

**Comment by svmhdvn** [comment]
+1 on this

Use case: one monorepo that runs the same test suite on multiple machine architectures (e.g. amd64, aarch64, riscv64, etc). Ideally, we'd express the Pipeline as a parallel stage comprising multiple stages, each with their own junit test reports. It would be great if the UI showed test reports per-stage instead of all aggregated into one per-job/per-pipeline as it is now.

**Comment by kaddkaka** [comment]
Is it possible to contribute to this? Does anyone have any insight in the current implementation and pointers before such an effort is started?

@uhafner have you been working on this or similar feature? How difficult do you think it would be to support multiple results file?

**Comment by uhafner** [comment]
> [@uhafner](https://github.com/uhafner) have you been working on this or similar feature? How difficult do you think it would be to support multiple results file?

No, I am not a contributor to this plugin. In the warnings and coverage plugins it was not difficult to implement it. The main difference is that the plugins report the results in multiple actions and not in a single one. But nowadays with the new Run UI it might not be necessary to use multiple actions? It might be sufficient to tag those results and show the results by the tag afterwards. See https://github.com/jenkinsci/warnings-ng-plugin/pull/2150 for the new UI that shows all warnings in a single view now. 

<img width="1424" height="769" alt="Image" src="https://github.com/user-attachments/assets/1a711c6c-164d-4143-8a08-bf03d68f1864" />

Since the JUnit plugin also has an implementation of the new Run UI it would be an interesting experiment to introduce the side bar for the JUnit plugin as well. 

<img width="1416" height="644" alt="Image" src="https://github.com/user-attachments/assets/1978fc25-3a8b-470e-8f07-0a39819d7e54" />

Internally this requires to change the storage of the results, so that the results can be filtered by a tag or something similar. I haven't looked at the code, how the tests are stored. Maybe we need to change a field from a single report to a list of reports. I am using such approach in my quality monitor, that shows the test results by a tag: https://github.com/jenkinsci/analysis-model/pull/1549#issuecomment-4857004357.

**Comment by kaddkaka** [comment]
I feel doubtful when I hear the word "filter" - I would prefer to never combine the test results into a total file. The reason for this being that it is extremely slow. Filtering through a million test results will not be fast.

A "list of reports" as you also mention sound much more appropriate! 🤗
